### PR TITLE
feat(recommendations): Provide a plug-in for recommendations

### PR DESCRIPTION
**Pull request by benoitf**
### What does this PR do?
Provide a plug-in for recommendations

1. Open a workspace without plug-ins: provide featured plug-ins and ask to reload workspace
2. Open a workspace containing a devfile with plug-ins: suggest it (but not prompting restart)

3. When opening file: Inform that some plug-ins may exist  (only if `extensions.openFileRecommendations` is set to true in devfile)

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/18274

### How to test this PR?
Starts a new workspace with this plug-in and a devfile without plug-ins but with java or go project.
--> should prompt to restart workspace and then have intellisense

Try to open some files like go or python: should suggest that existing plug-ins may be there (`extensions.openFileRecommendations` is set to true in devfile attributes)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [x] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [x] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [x] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=next

### CODE REVIEW

**Comment by benoitf** [comment]
[crw-ci-test]

**Comment by benoitf** [comment]
[crw-ci-test --rebuild]

File: plugins/recommendations-plugin/package.json
Comment (sunix) [inline plugins/recommendations-plugin/package.json:2]: Could that be the same name everywhere ?
here `recommendation-plugin` but the folder is `recommendationS-plugin`
Also most of our plugins name start with `@eclipse-che/`

**Review by sunix — CHANGES_REQUESTED** [review]

**Comment by sunix** [comment]
for che-theia, it recommended the `Java` and `python` plugin ... don't know if this is expected.

**Comment by sunix** [comment]
I have push the plugin to a registry to simplify testing, this is how you could test it (Taking a devfile, removing plugins and adding recommendation plugin):
```yaml
apiVersion: 1.0.0
metadata:
  name: golang-sau7u
projects:
  - name: example
    source:
      location: 'https://github.com/golang/example.git'
      type: git
    clonePath: src/github.com/golang/example/
  - name: golang-echo-realworld-example-app
    source:
      location: 'https://github.com/xesina/golang-echo-realworld-example-app.git'
      type: git
      branch: master
    clonePath: src/github.com/xesina/golang-echo-realworld-example-app
components:
  - id: redhat-developer/recommendations-plugin/latest
    type: chePlugin
    registryUrl: 'https://sunix-che-plugin-registry-recomandation-plugin-rqkin.surge.sh/v3/'
  - mountSources: true
    endpoints:
      - name: 8080-tcp
        port: 8080
    memoryLimit: 2Gi
    type: dockerimage
    image: 'quay.io/eclipse/che-golang-1.14:7.24.0'
    alias: go-cli
    env:
      - value: $(CHE_PROJECTS_ROOT)
        name: GOPATH
      - value: /tmp/.cache
        name: GOCACHE
commands:
  - name: 1.1 Run outyet
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/src/github.com/golang/example/outyet'
        type: exec
        command: go get -d && go run main.go
        component: go-cli
  - name: 1.2 Stop outyet
    actions:
      - type: exec
        command: kill $(pidof go)
        component: go-cli
  - name: 1.3 Test outyet
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/src/github.com/golang/example/outyet'
        type: exec
        command: go test
        component: go-cli
  - name: '2.1 xenisa :: install dependencies'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go mod download
        component: go-cli
  - name: '2.2 xenisa :: run'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go run main.go
        component: go-cli
  - name: '2.3 xenisa :: build'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go build
        component: go-cli
  - name: '2.4 xenisa :: test'
    actions:
      - workdir: '${GOPATH}/src/github.com/xesina/golang-echo-realworld-example-app'
        type: exec
        command: go test ./...
        component: go-cli
  - name: Run current file
    actions:
      - workdir: '${fileDirname}'
        type: exec
        command: 'go get -d && go run ${file}'
        component: go-cli
  - name: Debug current file
    actions:
      - referenceContent: |
          {
            "version": "0.2.0",
            "configurations": [
              {
                "name": "Debug current file",
                "type": "go",
                "request": "launch",
                "mode": "auto",
                "program": "${fileDirname}"
              }
            ]
          }
        type: vscode-launch


```

**Comment by sunix** [comment]
with the above plugin, i have 
Selection_178

But clicking `OK` doesn't do anything

**Comment by sunix** [comment]
I am going to make more test with the generated editor:
```yaml
  - reference: https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-916/simple/che-theia-editor.yaml
    type: cheEditor
```

**Comment by sunix** [comment]
Tested this devfile doesn't suggest anything:

```yaml
apiVersion: 1.0.0
metadata:
  name: java-web-spring-9op03
projects:
  - name: java-web-spring
    source:
      location: 'https://github.com/spring-projects/spring-petclinic.git'
      type: git
      branch: main
components:
  - type: cheEditor
    reference: 'https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-916/simple/che-theia-editor.yaml'
  - mountSources: true
    endpoints:
      - name: 8080-tcp
        port: 8080
      - attributes:
          public: 'false'
        name: debug
        port: 5005
    memoryLimit: 768Mi
    type: dockerimage
    volumes:
      - name: m2
        containerPath: /home/user/.m2
    image: 'quay.io/eclipse/che-java11-maven:7.24.0'
    alias: tools
    env:
      - value: ''
        name: MAVEN_CONFIG
      - value: '-XX:MaxRAMPercentage=50.0 -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xms20m -Djava.security.egd=file:/dev/./urandom -Duser.home=/home/user'
        name: JAVA_OPTS
      - value: $(JAVA_OPTS)
        name: MAVEN_OPTS
commands:
  - name: maven build
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/java-web-spring'
        type: exec
        command: mvn clean install
        component: tools
  - name: run webapp
    actions:
      - workdir: '${CHE_PROJECTS_ROOT}/java-web-spring'
        type: exec
        command: |
          java -jar -Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=n,address=5005 \
          target/*.jar
        component: tools
  - name: Debug remote java application
    actions:
      - referenceContent: |
          {
          "version": "0.2.0",
          "configurations": [
            {
              "type": "java",
              "name": "Debug (Attach) - Remote",
              "request": "attach",
              "hostName": "localhost",
              "port": 5005
            }]
          }
        type: vscode-launch

```

**Comment by sunix** [comment]
When opening a file, it is saying that some plugins may exist but it is just a message. Why not having a button to install the plugins? 
Selection_178

File: plugins/recommendations-plugin/src/fetch/featured-fetcher.ts
Comment (tsmaeder) [reply plugins/recommendations-plugin/src/fetch/featured-fetcher.ts:19]: I looked at the referenced gist and it's very incomplete right now.

**Review by tsmaeder — DISMISSED** [review]
What I like:
* well commented
* the code as written seems reasonable
What I like less:
* the code would profit from paying more attention to naming: code is written once, but read many times, and mostly by other people
* I don't think this feature is quite finished nor necessary right now. I'll comment more on the main PR page.

**Comment by tsmaeder** [comment]
There are a couple of issues with this feature as it stands now:

1. It requires us to maintain more files by hand: both the featured.json and 'https://raw.githubusercontent.com/benoitf/my-che-repository/master/v3/recommendations/language. We are still not done with the conversion from meta.yaml files to a central description that would describe single vscode extensions instead of che plugins. Who is going to maintain those assets?
IMO, we should first get rid of the one plugin, multiple vsx's relationship (make the sidecar structure an implementation detail) before adding features depending on a flawed model.
2. I think the approach is too much centered on "languages". I would like to propose the following model: The set of plugins we have in the workspace are activated by certain activation events. I would like to extend the universe of activated plugins to the ones that are "recommended", even if they are not in the workspace right now. We should hook into the regular theia activation mechanisms to achieve that (for example, "onDebug"). I believe that would be much simpler to achieve in an extension rather than a plugin. Ideally, we the recommendation mechanism would also work upstream (and be extended downstream).
3. I don't think recommendations are all that useful right now: our model is based on using the "stacks" that are predefined and where plugins already are defined in the devfile. With the advent of devfile 2.0 that might change, but right now, I don't see a pressing need.

**Comment by benoitf** [comment]
> There are a couple of issues with this feature as it stands now:
> It requires us to maintain more files by hand: both the featured.json and 'https://raw.githubusercontent.com/benoitf/my-che-> repository/master/v3/recommendations/language. We are still not done with the conversion from meta.yaml files to a central > description that would describe single vscode extensions instead of che plugins. Who is going to maintain those assets?
>IMO, we should first get rid of the one plugin, multiple vsx's relationship (make the sidecar structure an implementation detail) before adding features depending on a flawed model.

It will not be maintained by hand. It's automatically computed from vsix metadata.
I opened PR for early review (will not merged with my gist) but everything is generated by plugin-registry itself

> I think the approach is too much centered on "languages". I would like to propose the following model: The set of plugins we > have in the workspace are activated by certain activation events. I would like to extend the universe of activated plugins to
> the ones that are "recommended", even if they are not in the workspace right now. We should hook into the regular theia
> activation mechanisms to achieve that (for example, "onDebug"). I believe that would be much simpler to achieve in an
> extension rather than a plugin. Ideally, we the recommendation mechanism would also work upstream (and be extended
> downstream).

It's by design that it has been first using language first. But another trigger would be on workspaceContains as well metadata.
It's in che-theia because it's based on what is provided by the plug-in registry. (there is no metadata on open-vsx)
Also it's to suggest what user don't have.

> I don't think recommendations are all that useful right now: our model is based on using the "stacks" that are predefined
> and where plugins already are defined in the devfile. With the advent of devfile 2.0 that might change, but right now, I don't > see a pressing need.

I think it It's useful when you start from a given repository that don't have any devfile (because most people start from a project they want to work on and without having devfiles)
Let say I import petclinic java, it's very helpful to have java plug-in automatically enabled

**Comment by ericwill** [comment]
I think this is a good start to get some basic recommendation system implemented. It's something nice to implement on general principle, and it's important for the upcoming devfile 2.0 story. So from my POV, let's merge this (after comments are fixed) and iterate over it to improve it. I do like the idea of using activation events to trigger recommendations, and I think that's something that should be explored in future PR.

File: plugins/recommendations-plugin/package.json
Comment (benoitf) [reply plugins/recommendations-plugin/package.json]: it has been fixed

File: plugins/recommendations-plugin/src/analyzer/vscode-current-plugins-languages.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/analyzer/vscode-current-plugins-languages.ts]: fixed

File: plugins/recommendations-plugin/src/analyzer/vscode-current-plugins.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/analyzer/vscode-current-plugins.ts]: fixed

File: plugins/recommendations-plugin/src/fetch/featured-fetcher.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/fetch/featured-fetcher.ts]: it was to get early feedback, data is coming from plug-in registry. fixed

File: plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts]: it was to get early feedback, data is coming from plug-in registry. fixed

File: plugins/recommendations-plugin/src/logic/feature-plugin-logic-request.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/logic/feature-plugin-logic-request.ts]: used "strategy" in name

File: plugins/recommendations-plugin/src/analyzer/vscode-current-plugins.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/analyzer/vscode-current-plugins.ts]: fixed

Comment (benoitf) [reply plugins/recommendations-plugin/src/analyzer/vscode-current-plugins.ts]: fixed

File: plugins/recommendations-plugin/src/devfile/devfile-handler.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:21]: che server does not know about it, only che theia
can't consume it from somewhery

File: plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts]: fixed

File: plugins/recommendations-plugin/src/devfile/devfile-handler.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:102]: yes, but random

File: plugins/recommendations-plugin/src/fetch/featured-contribute-language.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/fetch/featured-contribute-language.ts:20]: I've added comments.
It's coming straight from language definition of package.json so not renaming t

File: plugins/recommendations-plugin/src/find/find-file-extensions.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/find/find-file-extensions.ts:16]: there is `vscode.workspace.findFiles` but it was not providing what I wanted, will investigate in a follow-up

File: plugins/recommendations-plugin/src/find/find-file-extensions.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/find/find-file-extensions.ts:30]: I kept findInFolder as we're inside a class named FindFIleExtensions so it's looking like duplicate to call
`findFileExtensions.findFileExtensionsInFolder` vs `findFileExtensions.findInFolder`

File: plugins/recommendations-plugin/src/plugin/recommendation-plugin.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/plugin/recommendation-plugin.ts]: using templates.
removed

File: plugins/recommendations-plugin/package.json
Comment (benoitf) [reply plugins/recommendations-plugin/package.json]: @sunix I fixed it 13 days ago so could you review it again ?
You required changes for this bad convention, and I fixed it a couple of weeks ago

**Comment by ericwill** [comment]
@tsmaeder @sunix are there any blocking concerns from your side? Or can we merge this and improve it.

**Comment by ericwill** [comment]
@benoitf this one needs a rebase

File: plugins/recommendations-plugin/src/devfile/devfile-handler.ts
Comment (tsmaeder) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:21]: Needs documentation, in that case.

File: plugins/recommendations-plugin/src/devfile/devfile-handler.ts
Comment (tsmaeder) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:102]: Let's open a bug and reference it in a comment, then.

File: plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts
Comment (tsmaeder) [reply plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts:19]: I still see the `benoitf` repository above.

File: plugins/recommendations-plugin/src/devfile/devfile-handler.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:21]: @tsmaeder it is planned to have documentation in che-docs. Are you talking about a different code comment or it's user documentation ?

File: plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts
Comment (benoitf) [reply plugins/recommendations-plugin/src/fetch/plugins-per-language-fetcher.ts]: I think I didn't rebased the correct branch. I've rebased with the correct branch now.

**Comment by benoitf** [comment]
I've rebased using the correct branch now of my fork. I think my latest changes were not reflected in the previous rebase.

File: plugins/recommendations-plugin/src/devfile/devfile-handler.ts
Comment (tsmaeder) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:21]: No, I mean user documentation. I thought we wanted to get back to a state where we need a companion docs PR before merge? Wouldn't this be a good time?

Comment (benoitf) [reply plugins/recommendations-plugin/src/devfile/devfile-handler.ts:21]: it won't be merged without user documentation as checkboxes are not checked IN PR checklist

**Comment by tsmaeder** [comment]
@benoitf there are still a couple of comments which are not addressed. You don't have to agree with me, but I'd like an answer.

**Comment by benoitf** [comment]
[crw-ci-test --rebuild]

**Comment by dmytro-ndp** [comment]
[crw-ci-test --rebuild]

**Comment by sunix** [comment]
Hello, testing as a user with this devfile:
```yaml
apiVersion: 1.0.0
metadata:
  generateName: java-maven-
projects:
  - name: console-java-simple
    source:
      type: git
      location: 'https://github.com/che-samples/console-java-simple.git'
      branch: java1.11
components:
  - reference: https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-916/simple/che-theia-editor.yaml
    type: cheEditor
```

So it doesn't do anything at startup of the workspace. Then when I open a Java file, it suggests me to install the plugin. But finally it is not updating the devfile.

**Comment by sunix** [comment]
For the installation of recommanded plugins when opening the java file (Could be done in another iteration):
- would be nice to tell the user to restart the workspace
- would be nice to show a list (not json content) of plugins to install
- would be nice to be able to choose the recommended plugins to install (at the moment, install java8 and java11 and quarkus11, etc ....) or install only one!

**Comment by benoitf** [comment]
[crw-ci-test --rebuild]

**Comment by benoitf** [comment]
> Hello, testing as a user with this devfile:
> So it doesn't do anything at startup of the workspace. Then when I open a Java file, installing the recommended plugins works.

there was a dependency issue on workspace-plugin (now it's marked as explicit dependency in package.json) so some events were not triggered. and there was a glitch in featured.json file.

should be fixed now

image



> For the installation of recommended plugins when opening the java file (Could be done in another iteration):
> would be nice to tell the user to restart the workspace

Should be implemented now


> would be nice to show a list (not json content) of plugins to install

I think it's a comma separated list of pluginId ?
I agree it's not the best and UX team has been involved few weeks ago to provide enhancements (for all user interactions) and will provide feedback

> would be nice to be able to choose the recommended plugins to install (at the moment, install java8 and java11 and quarkus11, etc ....) or install only one!

I tried but I faced https://github.com/eclipse-theia/theia/issues/5673 (Theia is not handling the `canPickMany` option in QuickPick)

**Comment by sunix** [comment]
Last changes:
- Fix problem no startup recommendation when starting a devfile with just a java project and the che-theia editor.
- We agree to Florent to disable the recommendation when a file is open. As it is displaying

**Review by sunix — APPROVED** [review]
LGTM

**Comment by sunix** [comment]
Restart is not working but covered with https://github.com/eclipse/che/issues/19174

**Comment by ericwill** [comment]
@tsmaeder any comments left from your side? Otherwise it's good to merge IMO

**Comment by tsmaeder** [comment]
Nope, all good.

**Pull request merged**
